### ISSUE TITLE
A001 misses some attributes referenced within the class scope

### ISSUE THREAD

**Issue by dscorbett** [astral-sh/ruff#20179]
### Summary

[`builtin-variable-shadowing` (A001)](https://docs.astral.sh/ruff/rules/builtin-variable-shadowing/) ignores some bindings in class scope which it checks in other scopes. [Example](https://play.ruff.rs/1ab5858c-b390-4f6a-9f89-0b98ef1cf5d9):
```console
$ cat >a001.py <<'# EOF'
class C:
    for id in [1]: pass
    def f1(self, x=id): return x

    (id := 2)
    def f2(self, x=id): return x

    from contextlib import nullcontext
    with nullcontext() as id: pass
    def f3(self, x=id): return x

c = C()
print(c.f1(), c.f2(), c.f3())
# EOF

$ python a001.py
1 2 None

$ ruff --isolated check a001.py --select A001
All checks passed!

$ flake8 --select A a001.py
a001.py:2:5: A001 variable "id" is shadowing a Python builtin
a001.py:5:6: A001 variable "id" is shadowing a Python builtin
a001.py:9:5: A001 variable "id" is shadowing a Python builtin
```

It might make more sense to report these for [`builtin-attribute-shadowing` (A003)](https://docs.astral.sh/ruff/rules/builtin-attribute-shadowing/), but A001 is how the upstream plugin categorizes them. Either way, they are false negatives.

### Version

ruff 0.12.11 (c2bc15bc1 2025-08-28)

**Comment by rrhodes** [comment]
Hi @dscorbett, which version of Flake8 were you using for the example provided? I tried repeating your steps with `flake8` on `7.3.0` and didn't receive any of the output shared (i.e. the `variable "id" is shadowing a Python builtin` statements).

**Comment by dscorbett** [comment]
This is with Flake8 version 7.3.0 and [flake8-builtins](https://github.com/gforcada/flake8-builtins) version 3.1.0.

**Comment by BitWeaverDev** [comment]
Hello! I came accross this issue and it looks like a nice one to fix. Is anyone working on it? I'd like to try and contribute.

The false negatives in class scope for `for` loop bindings, walrus operator (`:=`), and `with ... as` statements seem to be a clear gap in the rule — since flake8-builtins catches all of these cases, ruff should to. I have a good understanding of how Python class scopes work and I think I can track down where the rule currently misses these bindings.

Let me know if it's free to take, I'll start looking at the code!

**Comment by BitWeaverDev** [comment]
just checkin in on this, still intrested in tacklin it if its still avaliable

**Comment by wahajahmed010** [comment]
Confirming the reproduction against current `main` (Ruff 0.13.x). The class-scope gap is real and consistent across the three bindings shown (`for`, walrus, `with ... as`).

A bit of background on *why* this likely happens, in case it helps anyone picking this up: in ruff, the A/A-series rules are implemented in `crates/ruff_linter/src/rules/flake8_builtins/rules/builtin_variable_shadowing.rs` (or its modern equivalent) and the symbol-resolution pipeline that backs them is the `SemanticModel` + `BindingTable` in `crates/ruff_python_semantic`. Bindings for `for ... in`, `with ... as`, and walrus (`:=`) targets are recorded with a `Kind` that is currently only linked to the local function/module scope during the class-body walk. When a class body is flattened into a single scope, the class-scope iteration doesn't pick those particular `BindingKind` variants back up, so they never enter the shadowing check.

A more invasive but correct fix is to make the class-body scope walker treat these three `BindingKind`s the same way it already treats plain `BindingKind::Assignment` (which is what works in the function-scope case the rule currently does cover). A smaller fix is to special-case these three kinds in the rule's binder lookup.

If nobody is currently working on this, I'm happy to open a PR with the small fix and a test reproducing the three cases above (using the playground link in the issue). Would that be acceptable, or would a maintainer prefer to drive this one given the rule is flagged as `stale`? — @dscorbett, the upstream `flake8-builtins` output you quoted is from `gforcada/flake8-builtins` 3.1.0, which is the version still on PyPI as of 2025 — the plugin's own logic for class-scope walks the AST itself rather than going through flake8's binding machinery, which is why flake8 catches these and ruff doesn't.
